DEVELOPMENT.md

```diff
@@ -179,3 +179,43 @@ Here is a short summary of the various commands and their primary function. Plea
 
     However, you cannot use `ibazel run` in this case. The file watcher is glitchy on running the tests
     when detecting changes. It shows '`a broken pipe`' in terminal. We need to terminate and restart the program manually.
+
+## Adding, Updating, or Removing Frontend Dependencies
+
+For the most part, frontend-specific third-party dependencies are hosted by
+[npm](https://docs.npmjs.com/about-npm) and managed by
+[yarn](https://classic.yarnpkg.com/). Some of our build-time dependencies are
+managed by [bazel](https://bazel.build/).
+
+The source of truth for `yarn` is a combination of the `package.json` and
+`yarn.lock` files. `package.json` is maintained by us developers and describes
+direct dependencies while `yarn.lock` is maintained by the system and describes
+the entire tree of transitive dependencies. In reality both files are often
+edited automatically by calls to `yarn` and `package.json` only sometimes needs
+to be edited by hand.
+
+1.  Install yarn.
+
+2.  Add or modify an entry in the `[dependencies]` or `[devDependencies]`
+    section of `package.json`. You can do this manually but often it's preferred
+    to use `yarn` from the command line:
+    * `yarn add`
+      * https://classic.yarnpkg.com/lang/en/docs/cli/add/
+    * `yarn upgrade`
+      * See: https://classic.yarnpkg.com/lang/en/docs/cli/upgrade/
+      * Alternatively, use `yarn upgrade-interactive`, possibly with the
+        `--latest` flag to use an interactive shell tool to hand pick
+        dependencies to upgrade.
+      * See: https://classic.yarnpkg.com/lang/en/docs/cli/upgrade-interactive/
+    * `yarn remove`
+      * https://classic.yarnpkg.com/lang/en/docs/cli/remove/
+
+3.  Run `yarn run yarn-deduplicate`.
+
+4.  Cross reference your updates with the bazel `WORKSPACE` file and determine if any
```

Comment:
This is a crucial piece for some of the upgrades. I think a bit more explanation could be useful. Maybe a link to [go/tensorboard-tf-mirror](http://goto.google.com/tensorboard-tf-mirror) once that is working again? That being said this is more than I had to go off of.

Reply:
Added mirror information.

I don't have any more detailed information to provide because I have little experience with this part. The esbuild update required some modifications to WORKSPACE but not in any way that was worth documenting in detail. If you have ideas on how to further improve this doc I will happily review your PR.